python-sdk/nuscenes/scripts/nuscenes_kitti_roundtrip.py

```diff
@@ -0,0 +1,430 @@
+# nuScenes dev-kit.
+# Code written by Holger Caesar, 2019.
+# Licensed under the Creative Commons [see licence.txt]
+
+
+"""
+This script converts nuScenes data to KITTI format and vice versa.
+It is used for compatibility with software that uses KITTI-style annotations.
+
+We do not encourage this, as:
+- KITTI has only front-facing cameras, whereas nuTonomy has a 360 degree horizontal fov.
+- KITTI has no radar data.
+- The nuScenes database format is more modular.
+- KITTI fields like occluded and truncated cannot be exactly reproduced from nuScenes data.
+- We don't specify the KITTI imu_to_velo_kitti projection in this code base.
+- KITTI has different categories.
+
+To validate the outcome we do the following:
+- Test that the roundtrip conversion (nuScenes -> KITTI -> nuScenes and vice versa) leads to the original outcome.
+- Visualize KITTI examples.
+
+The two main methods here are:
+- `nuscenes_roundtrip`: Converts nuScenes to KITTI files, converts them back and checks equivalence.
+                        The resulting KITTI files can be used in other codebases designed for KITTI.
+- `kitti_roundtrip`:    Converts KITTI boxes to nuScenes boxes and checks equivalence.
+                        This script exists only for instructional purposes and does not create a valid nuScenes-style
+                        dataset.
+"""
+import os
+import json
+from typing import List, Tuple
+import argparse
+
+import numpy as np
+from pyquaternion import Quaternion
+from PIL import Image
+
+from nuscenes.nuscenes import NuScenes
+from nuscenes.utils.geometry_utils import transform_matrix
+from nuscenes.utils.geometry_utils import BoxVisibility
+from nuscenes.utils.data_classes import LidarPointCloud
+from nuscenes.utils.splits import create_splits_logs
+from nuscenes.utils.kitti import KittiDB
+
+
+def nuscenes_roundtrip(nusc: NuScenes,
+                       splits: Tuple[str, ...],
+                       kitti_fake_dir: str) -> None:
+    """
+    Check that boxes can be converted from nuScenes to KITTI and back.
+    :param nusc: NuScenes instance.
+    :param splits: The list of relevant splits (e.g. train, val).
+    :param kitti_fake_dir: Where to write the KITTI-style annotations.
+    """
+    # Convert to KITTI files.
+    nuscenes_to_kitti_file(nusc=nusc, splits=splits, kitti_fake_dir=kitti_fake_dir)
+
+    # Check if the input can be reconstructed.
+    kitti_file_to_nuscenes_check(nusc=nusc, splits=splits, kitti_fake_dir=kitti_fake_dir)
+
+    print('Passed nuScenes roundtrip check!')
+
+
+def kitti_roundtrip(kitti_dir: str,
```

Comment:
So this method doesn't save anything to disk? So we don't provide a way to go from KITTI to the nuScenes format (on disk)?

Reply:
Entire script restructured now. Going back to nuScenes GT format is no longer supported.